manual: clarify documentation of C macros on integers

nojb opened the pull request:
It was pointed out in #13844 that the documentation of the C macros on integers (`Val_long`, `Long_val`, etc) could be clarified. This is the purpose of this PR.

Comments welcome.

Fixes #13844

nojb committed:
manual: clarify documentation of C macros on integers

gasche commented on manual/src/cmds/intf-c.etex:
```diff
@@ -702,12 +702,20 @@ type) corresponds to the "Some" constructor.
-\item "Val_bool("\var{x}")" returns the OCaml boolean representing the
-truth value of the C integer \var{x}.
+\item "Val_long("\var{n}")" returns the value encoding the C integer \var{n} of
+type "intnat". Despite its name, note that "intnat" is not necessarily the same
```
There is no explicit documentation of `intnat` in this file. Could you maybe document this type on its own, before the list of operations?

gasche commented on manual/src/cmds/intf-c.etex:
```diff
@@ -702,12 +702,20 @@ type) corresponds to the "Some" constructor.
+\item "Long_val("\var{v}")" returns the C integer of type "intnat" encoded in a
+value \var{v}.
+\item "Val_int("\var{i}")" returns the value encoding of the C integer \var{i}
+of type "int".  It coincides with "Val_long("\var{i}")".
```
I'm not fond of the "it coincides with ..." comments, which add weight to the documentation (sometimes repeating the implementation) and I'm not sure there are matching benefits.

gasche commented on manual/src/cmds/intf-c.etex:
```diff
@@ -702,12 +702,20 @@ type) corresponds to the "Some" constructor.
+coincides with "(int)Val_long("\var{v}")".
+\item "Val_bool("\var{x}")" returns the OCaml boolean representing the truth
+value of the C integer \var{x} (where 0 is interpreted as "false" and any other
+value as "true").
```
I wonder if we could try giving these operations as C prototypes (eg. `int Bool_val(value)`). This would be clearer in several cases. But I'm not sure what type to give to certain macros, for example `#define Val_bool(x) Val_int((x) != 0)`, I guess that there is no principal type.

gasche reviewed (commented):
I made some comments. My high-level criticism: we keep adding extra details that are useful for experts, and we risk losing the crispiness of the documentation that makes it pleasant to read. I think that ideally we would try to (1) compress the more-detailed form into something short enough again, if we figure how to do this without losing more information, or (2) moving some extra details outside the list as a side paragraph, before the list (preliminary notions, eg. `intnat`) or after (warnings, precisions, etc.).

nojb committed:
Fewer details

nojb commented:
Thanks @gasche for the review. I tried to take your comments into account, as follows:
- Reverted to the original formulation of the manual, which is a bit less precise but lighter on the reader; with one change, namely I replaced "long int" by "intnat" (this is a bug fix);
- Added a small blurb in that same section about what "intnat" is;
- (Unrelated) removed an unnecessary duplication in the description of "Val_emptylist"

dbuenzli commented:
Personally I don't find much of an improvement here (sorry!). Here are suggestions:

1. The convention `Val_XXX` `XXX_val` could be explained along the lines @gasche did in [#13844](#13844). I.e. `XXX` mentions C types not OCaml types.
2. I think it should be mentioned in the description of `Int_val` that since it casts to `int` it might silently truncate an OCaml integer on 64-bit platforms.

nojb committed:
Add note about Val_int truncation

nojb commented:
> Personally I don't find much of an improvement here (sorry!). Here are suggestions:
> 
> 1. The convention `Val_XXX` `XXX_val` could be explained along the lines @gasche did in [#13844](#13844). I.e. `XXX` mentions C types not OCaml types.

I am not sure how general this comment would, since this does not apply for all `Val_xxx` macros (eg `Val_none`, `Val_emptylist`, `Val_true`, etc). It really only refers to `Val_long` and `Val_int` as far as I can see (am I forgetting any other?). So perhaps just clarifying those two operations is better.

> 2. I think it should be mentioned in the description of `Int_val` that since it casts to `int` it might silently truncate an OCaml integer on 64-bit platforms.

I added a note to this effect. Thanks.

dbuenzli commented:
> It really only refers to `Val_long` and `Val_int` as far as I can see (am I forgetting any other?). So perhaps just clarifying those two operations is better.

Perhaps. Though @gasche was hard at trying to convince me in #13844 that all this made sense for others too and it's true that in my bindings I usualy use `c_struct_val` for getting these pointers out of custom blocks and `c_type_val` and `Val_c_type` when it can be defined directly (e.g. for enums).

I guess it's just my brain who wanted to see `intnat`s behind `Val_int` and `Int_val`. But I have to say that the fact that `Val_int` is defined as `Val_long` did not help during the past 20 years. If `Val_int` had first cast its argument to `int` it would have all made sense and my wrong interpretation of these macros would likely have surfaced much earlier. I guess that one of these days I will need to have a serious review of `git grep Int_val` over my code bases.

gasche reviewed (approved):
Was reminded of this PR by the triaging meeting. It's fine now! Thanks everyone.
